HIP/hip-679.md

```diff
@@ -0,0 +1,215 @@
+
+##### Compression zstd vs gz:
+- zstd has a better compression ration that gzip.
+- (zstd:gzip)
+  - Faster compression - About a 5:1  ratio.
+  - Faster decompression - About a 3:1 ratio.
+
+##### Algorithm:
+- Read stream files from current location 
+    ``` 
+        accountBalances/balance0.0.3/ 
+        eventsStreams/events_0.0.3/     
+        recordstreams/record0.0.3/ 
+    ```
+    
+- The new path for aggregated files would be:
+    
+    ``` 
+        {network}/{shard}/record/
+        {network}/{shard}/balance/
+        {network}/{shard}/event/
+    ```
+- The lambda will decompress the file, if it is compressed.
+- Calculate a hash of the file contents.
+- Maintain an in-memory map which has the hashes for the current file being processed, to remove duplicate files with the same content.
+- Compress the file using zst if it not a signature file.
+- Monitor the size/time of the potential aggregated file.
+- For aggregation:
+  - Calculate the node Id based on the current node account id. NodeId = (number portion of) nodeAccountId - 3
+  ``` Node for balance0.0.4 would be 4-3 = 1  ```
+  - The new file structure would be as follows:
+    Example:
+    ```
+    >tar -tf 2022-07-13T08_46.tar | sort  
+
+    2022-07-13T08_46/
+    2022-07-13T08_46/0/
+    2022-07-13T08_46/0/2022-07-13T08_46_08.041986003Z.rcd.zst
+    2022-07-13T08_46/0/2022-07-13T08_46_08.041986003Z.rcd_sig
+    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z.rcd.zst
+    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z.rcd_sig
+    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z_01.rcd.zst
+    2022-07-13T08_46/1/
+    2022-07-13T08_46/1/2022-07-13T08_46_08.041986003Z.rcd.zst
+    2022-07-13T08_46/1/2022-07-13T08_46_08.041986003Z.rcd_sig
+    
+     ```
+     The above compressed tar file includes the following files in node id based folders:
+     - Record stream files (.rcd.zst)
+     - Signature files (.rcd_sig)
+     - Sidecar files (_01.rcd.zst)
+     
+     NOTE: File `2022-07-13T08_46_08.041986003Z.rcd.zst` appears in two separate node folders. These are files with distinct content and when downloaded by the mirror node one of the two file will be selected based on the consensus majority.
+     
+     The aggregation in the new files would be size/time based, as will be determined by performance testing.
+- Once the limit has been reached, aggregate the file and write it to the new bucket location. 
+- Delete the existing files that have been read from the current bucket location.
+
+After a significant amount of time has passed after launching Phase three, we will remove the properties in mirror node to default the downloading from the node ID based path.
+
+### Mirror Node Implications
+The existing mirror nodes would need to implement phase one.
+
+### Dev ops Implications
+After the completion of phase one and having being notified, dev ops will need to implement phase two.
```

Comment:
DevOps will be responsible for executing the rollout of this phase, but I don't believe that is the right assignment for overall implementation. This change will require updates to consensus node software, which would be implemented by services team. This could also require updates to mirror.py, which would be implemented by engineering as well (currently RE). If there are changes required to the deployment manifest (e.g. docker-compose changes) this part would currently go to DevOps to implement

Reply:
There should not be any consensus nodes changes unless we decide to have consensus nodes compress using zstd instead of gzip as Jasper has suggested. There will be mirror.py changes but that has only recently changed responsibilities from DevOps back to engineering in the past week or so, as I understand it. So we can update the this section to reflect that.

Reply:
Though, depending upon whether it's a config or code change to mirror.py it might be either DevOps or engineering.

Reply:
I'm not sure what the update needs to be, but this statement still needs to be addressed

Reply:
My hope is longer term we will upload directly from services app rather than depending on external script. We can compress the stream as we send it and it can never touch the disk.

I did some experiments with zstd it can offer a nice increase in compression. It takes a little more CPU and a bit longer compared to Gzip to get there. If we build a dictionary that is reused for every file then it is less CPU and faster than Gzip and more compression. Just a bit more complex to code and manage. 

It would be nice to go from GZ to ZStd but not critical in short term

Reply:
Can this be resolved as of now?